src/main/java/com/gotocompany/depot/maxcompute/client/MaxComputeClient.java

```diff
@@ -0,0 +1,155 @@
+                .getSchema();
+        checkPartitionPrecondition(oldSchema);
+        List<String> schemaDifferenceSql = SchemaDifferenceUtils.getSchemaDifferenceSql(oldSchema, tableSchema, datasetName, tableName);
+        for (String sql : schemaDifferenceSql) {
+            Instance instance = execute(sql);
+            if (!instance.isSuccessful()) {
+                instrumentation.logError("Failed to execute SQL: " + sql);
+                String errorMessage = instance.getRawTaskResults().get(0).getResult().getString();
+                throw new MaxComputeTableOperationException(String.format("Failed to update table schema with reason: %s", errorMessage));
+            }
+        }
+    }
+
+    private void checkPartitionPrecondition(TableSchema oldSchema) {
+        if (maxComputeSinkConfig.isTablePartitioningEnabled() && oldSchema.getPartitionColumns().isEmpty()) {
+            throw new MaxComputeTableOperationException("Updating non-partitioned table to partitioned table is not supported");
+        }
+        if (maxComputeSinkConfig.isTablePartitioningEnabled()) {
+            String currentPartitionColumnKey = oldSchema.getPartitionColumns()
+                    .stream()
+                    .findFirst()
+                    .map(Column::getName)
+                    .orElse(null);
+            if (!Objects.equals(maxComputeSinkConfig.getTablePartitionColumnName(), currentPartitionColumnKey)) {
+                throw new MaxComputeTableOperationException("Changing partition column is not supported");
+            }
+        }
+    }
+
+    public void insert(List<RecordWrapper> recordWrappers) throws TunnelException, IOException {
+        insertManager.insert(recordWrappers);
+    }
+
+    private Odps initializeOdps() {
+        Account account = new AliyunAccount(maxComputeSinkConfig.getMaxComputeAccessId(), maxComputeSinkConfig.getMaxComputeAccessKey());
+        Odps odpsClient = new Odps(account);
+        odpsClient.setDefaultProject(maxComputeSinkConfig.getMaxComputeProjectId());
+        odpsClient.setEndpoint(maxComputeSinkConfig.getMaxComputeOdpsUrl());
+        odpsClient.setCurrentSchema(maxComputeSinkConfig.getMaxComputeSchema());
+        odpsClient.setGlobalSettings(getGlobalSettings());
+        return odpsClient;
+    }
+
+    private InsertManager initializeInsertManager() {
+        return InsertManagerFactory.createInsertManager(maxComputeSinkConfig, tableTunnel, instrumentation, maxComputeMetrics);
+    }
+
+    private Map<String, String> getGlobalSettings() {
+        Map<String, String> globalSettings = new HashMap<>();
+        globalSettings.put("setproject odps.schema.evolution.enable", "true");
+        globalSettings.put("odps.namespace.schema", "true");
+        return globalSettings;
+    }
+
+    public Instance execute(String sql) throws OdpsException {
+        log.info("Executing SQL: {}", sql);
+        return SQLTask.run(odps, sql);
+    }
```

Comment:
The execute method doesn't implement any circuit breaker pattern. We should add circuit breaker to prevent cascade failures when MaxCompute service is experiencing issues. Also missing proper error handling and retries.

Reply:
Could you please provide the example of the circuit breaker pattern for this context? This DDL operation wouldn't be frequently be executed anyway. 

I agree with the point of the error handling and retry mechanism, have implemented it in top of the method that called this